fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java

```diff
@@ -3505,130 +3497,265 @@ public void replayRecoverPartition(RecoverInfo info) {
+        Thread t = new Thread(() -> {
+            Map<Long, List<Long>> taskSignatures = new HashMap<>();
+            int numFinishedTasks;
+            int numSendedTasks = 0;
+            for (Partition partition : partitions) {
                 if (!countDownLatch.getStatus().ok()) {
-                    errMsg += " Error: " + countDownLatch.getStatus().getErrorMsg();
-                } else {
-                    List<Entry<Long, Long>> unfinishedMarks = countDownLatch.getLeftMarks();
-                    // only show at most 3 results
-                    List<Entry<Long, Long>> subList = unfinishedMarks.subList(0, Math.min(unfinishedMarks.size(), 3));
-                    if (!subList.isEmpty()) {
-                        errMsg += " Unfinished mark: " + Joiner.on(", ").join(subList);
+                    break;
+                }
+                List<CreateReplicaTask> tasks = buildCreateReplicaTasks(dbId, table, partition);
+                for (CreateReplicaTask task : tasks) {
+                    List<Long> signatures = taskSignatures.computeIfAbsent(task.getBackendId(), k -> new ArrayList<>());
+                    signatures.add(task.getSignature());
+                }
+                sendCreateReplicaTasks(tasks, countDownLatch);
+                numSendedTasks += tasks.size();
+                numFinishedTasks = numReplicas - (int) countDownLatch.getCount();
+                while (numSendedTasks - numFinishedTasks > 200 * numBackends) {
+                    try {
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
                     }
+                    numFinishedTasks = numReplicas - (int) countDownLatch.getCount();
                 }
-                LOG.warn(errMsg);
-                throw new DdlException(errMsg);
             }
+            if (countDownLatch.getStatus().ok()) {
+                taskSignatures.clear();
+            }
+            for (Map.Entry<Long, List<Long>> entry : taskSignatures.entrySet()) {
+                for (Long signature : entry.getValue()) {
+                    AgentTaskQueue.removeTask(entry.getKey(), TTaskType.CREATE, signature);
+                }
+            }
+        });
+        t.start();
+        try {
+            waitForFinished(countDownLatch, Math.min(timeout, maxTimeout));
+        } catch (Exception e) {
+            countDownLatch.countDownToZero(new Status(TStatusCode.UNKNOWN, e.getMessage()));
+            throw e;
+        }
+    }
 
-            if (index.getId() != baseIndexId) {
-                // add rollup index to partition
-                partition.createRollupIndex(index);
+    private List<CreateReplicaTask> buildCreateReplicaTasks(long dbId, OlapTable table, List<Partition> partitions) {
+        if (partitions.size() == 1) {
+            return buildCreateReplicaTasks(dbId, table, partitions.get(0));
+        }
```

Comment:
if think this branch can be removed.

Reply:
It was designed to avoid the `List` object creation and copy, but anyway, I removed this branch.